### PR TITLE
Fully Implement Nomic Embedding Model

### PR DESCRIPTION
**Pull request by ncylich**
# Fully Implement Nomic Embedding Model

## Summary
Complete end-to-end implementation of the Nomic BERT embedding model (nomic-embed-text-v2-moe) 
with full forward pass, Mixture-of-Experts (MoE) support, and comprehensive test coverage. 
Enables text embedding inference with bidirectional attention and sparse expert routing.

### 1. MoE-Specific Graph Operations
- **`index()`** - Tensor indexing along dimensions
- **`layernorm()`** - Layer normalization with weight/bias
- **`topk()`** - Top-K selection returning indices and values
- **`scatter_topk()`** - Scatter selected values into full tensor
- **Bidirectional attention** - `is_causal=false` flag for encoder models

### 2. Full Forward Pass Implementation
- **`build_attention()`** - Multi-head bidirectional attention with RoPE and bias
- **`build_standard_mlp()`** - Dense FFN layers with GELU activation
- **`build_moe_mlp()`** - Sparse expert routing with top-k gating:
  - Router softmax over expert logits
  - Top-K expert selection per token
  - Weighted combination of expert outputs
- **`build_transformer_block()`** - Pre-norm architecture with residual connections
- **`forward()`** - End-to-end embedding generation from token IDs

## Technical Highlights
- Bidirectional self-attention (no causal masking)
- Dynamic MoE layer selection (every N layers configurable)
- Proper precision handling (FP16/FP32) with casting for transposed MatMuls
- Expert output aggregation via scatter and weighted summation
- Mean pooling over sequence dimension for final embeddings

## Testing
New test suite (`tests/test_nomic_model.cpp`) validates:
- Forward pass execution with real tokenized input
- Cache guard enforcement (encoder models don't use KV cache)
- Attention and transformer block cache restrictions

## Files Changed
- `cactus/models/model_nomic.cpp` (+224/-11 lines) - Full implementation
- `tests/test_nomic_model.cpp` (+243 lines, new) - Test suite
- `cactus/graph/graph.h` (+23 lines) - New op types
- `cactus/graph/graph_builder.cpp` (+97 lines) - Builder methods
- `cactus/graph/graph_ops.cpp` (+282 lines) - Compute implementations
- `cactus/kernel/kernel_attention.cpp` (+15 lines) - Bidirectional support
- `cactus/engine/engine.h` (+6 lines) - MoE config params
- `tools/convert_hf.py` (+110/-22 lines) - Nomic BERT conversion

### CODE REVIEW

File: tests/test_nomic_model.cpp
Comment (HenryNdubuaku) [inline tests/test_nomic_model.cpp:1]: We can skip a separate test, test_engine already has a test_embedding, this should work within that setup, you can test that an then remove this.

File: .gitignore
Comment (HenryNdubuaku) [inline .gitignore:1]: This file changes seems like it is for a package you personally installed, so we dont quite need it in the repo

File: cactus/models/model_nomic.cpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+    auto gate_weights = gb->matmul(normalized_h, layer.mlp_router_layer_weight, true, backend);
+    auto gate_probs = gb->softmax(gate_weights);
+    
+    auto topk_indices_and_values = gb->topk(gate_probs, config_.num_top_experts);
+    auto topk_idx = gb->index(topk_indices_and_values, 0, 0);
+    auto topk_w = gb->index(topk_indices_and_values, 1, 0);
+    auto expert_token_weights_matrix = gb->scatter_topk(topk_idx, topk_w, num_experts);  // [E, N]
+
+    // Getting expert outputs
```
Comment (HenryNdubuaku) [inline cactus/models/model_nomic.cpp]: Are these comments necessary? one can easily inspect the profile.txt to see the dimensions right?

File: cactus/models/model_nomic.cpp [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    auto q_proj_4d = reshape_to_heads(q_proj);
+    auto k_proj_4d = reshape_to_heads(k_proj);
+    auto v_proj_4d = reshape_to_heads(v_proj);
+
+    if (config_.rope_theta > 0) {
+        q_proj_4d = gb->rope(q_proj_4d, config_.rope_theta, 0);
+        k_proj_4d = gb->rope(k_proj_4d, config_.rope_theta, 0);
+    }
+
+    auto attn_output_4d = gb->attention(q_proj_4d, k_proj_4d, v_proj_4d, attention_scale_, false);  // is_causal=false for bidirectional attention
```
Comment (HenryNdubuaku) [inline cactus/models/model_nomic.cpp]: Remove comments too
